nighthawk/source/common/utility.cc

```diff
@@ -34,21 +34,35 @@ Uri::Uri(const std::string& uri) : scheme_("http") {
   Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
   host_and_port_ = std::string(host);
   path_ = std::string(path);
-  const size_t colon_index = host_and_port_.find(':');
   const bool is_https = absl::StartsWith(uri, "https://");
   const size_t scheme_end = uri.find("://", 0);
-
   if (scheme_end != std::string::npos) {
     scheme_ = absl::AsciiStrToLower(uri.substr(0, scheme_end));
   }
 
+  size_t colon_index = std::string::npos;
+  bool in_ipv6_address = false;
+
+  for (size_t i = 0; i < host_and_port_.size(); i++) {
```

Comment:
Don't any of the existing Envoy address parsing utils help here? I'm surprised we're the first to be deconstructing an IPv6 address..

Reply:
If we need this, I suggest splitting out into a separate utility and adding tests..

Reply:
Well, there's `Utility::parseInternetAddressAndPort()` (which we now use in the client later on). But it expects a port, which we do not always have (the code is working to figure that out here). Splitted this out and added tests; also added some limited tests for hostname validation

Reply:
Hmm, this answers my earlier comment. Can't we just do a quick scan from RHS, check if there is a `:`, and if not, add a dummy port?